File: 1673. Find the Most Competitive Subsequence/Solution.py
from typing import List


class Solution:
    def mostCompetitive(self, nums: List[int], k: int) -> List[int]:
        stack = []
        for i, x in enumerate(nums):
            while stack and x < stack[-1] and len(nums) - i + len(stack) > k:
                stack.pop()
            if len(stack) < k:
                stack.append(x)
        return stack

File: 1673. Find the Most Competitive Subsequence/Solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> mostCompetitive(vector<int>& nums, int k)
    {
        vector<int> mono_stk;
        for (size_t i = 0; i < nums.size(); i++) {
            while (!mono_stk.empty() && nums[i] < mono_stk.back() && nums.size() - i + mono_stk.size() > k)
                mono_stk.pop_back();

            if (mono_stk.size() < k)
                mono_stk.emplace_back(nums[i]);
        }
        return mono_stk;
    }
};
